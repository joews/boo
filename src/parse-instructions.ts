import { readFileSync } from "fs";
import { join } from "path";
import { generate } from "pegjs";
import { Instruction } from "./types";

const parserSource = readFileSync(
  join(__dirname, "instructions.pegjs"),
  "utf8"
);
const parser = generate(parserSource);
const instructions = readFileSync(join(__dirname, "boo.instructions"), "utf8");

export default function read(): Instruction[] {
  return parse(instructions);
}

export function parse(source: string): Instruction[] {
  const instructions = parser.parse(source) as Instruction[];
  // console.log(`TRACE parse instructions: `, JSON.stringify(instructions));
  return instructions;
}
