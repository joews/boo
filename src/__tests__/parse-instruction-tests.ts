/* eslint-env jest */
import { parse, default as read } from '../parse-instructions'

function testParse (source: string) {
  test(source, () => {
    // console.log(JSON.stringify(parse(source)))
    expect(parse(source)).toMatchSnapshot()
  })
}

testParse('0x00 noop (): () -> ()')
testParse('0x10 iadd (): (int32, int32) -> (int32)')
testParse('0x20 jmp (label): () -> ()')

test('read all instructions', () => {
  const instructions = read()
  // console.log(JSON.stringify(instructions, null, 2))

  expect(instructions).toMatchSnapshot()
})
