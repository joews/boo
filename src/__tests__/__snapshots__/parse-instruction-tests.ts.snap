// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0x00 noop (): () -> () 1`] = `
Array [
  Object {
    "argTypes": Array [],
    "mnemonic": "noop",
    "opcode": "0x00",
    "operandTypes": Array [],
    "resultTypes": Array [],
  },
]
`;

exports[`0x10 iadd (): (int32, int32) -> (int32) 1`] = `
Array [
  Object {
    "argTypes": Array [],
    "mnemonic": "iadd",
    "opcode": "0x10",
    "operandTypes": Array [
      "int32",
      "int32",
    ],
    "resultTypes": Array [
      "int32",
    ],
  },
]
`;

exports[`0x20 jmp (label): () -> () 1`] = `
Array [
  Object {
    "argTypes": Array [
      "label",
    ],
    "mnemonic": "jmp",
    "opcode": "0x20",
    "operandTypes": Array [],
    "resultTypes": Array [],
  },
]
`;

exports[`read all instructions 1`] = `
Array [
  Object {
    "argTypes": Array [
      "int32",
    ],
    "mnemonic": "iconst",
    "opcode": "0x00",
    "operandTypes": Array [],
    "resultTypes": Array [
      "int32",
    ],
  },
  Object {
    "argTypes": Array [],
    "mnemonic": "iadd",
    "opcode": "0x01",
    "operandTypes": Array [
      "int32",
      "int32",
    ],
    "resultTypes": Array [
      "int32",
    ],
  },
  Object {
    "argTypes": Array [],
    "mnemonic": "print",
    "opcode": "0x02",
    "operandTypes": Array [
      "int32",
    ],
    "resultTypes": Array [],
  },
  Object {
    "argTypes": Array [
      "int32",
    ],
    "mnemonic": "gstore",
    "opcode": "0x03",
    "operandTypes": Array [
      "int32",
    ],
    "resultTypes": Array [],
  },
  Object {
    "argTypes": Array [
      "int32",
    ],
    "mnemonic": "gload",
    "opcode": "0x04",
    "operandTypes": Array [],
    "resultTypes": Array [
      "int32",
    ],
  },
  Object {
    "argTypes": Array [
      "label",
    ],
    "mnemonic": "jmp",
    "opcode": "0x05",
    "operandTypes": Array [],
    "resultTypes": Array [],
  },
  Object {
    "argTypes": Array [],
    "mnemonic": "halt",
    "opcode": "0x06",
    "operandTypes": Array [],
    "resultTypes": Array [],
  },
  Object {
    "argTypes": Array [
      "label",
    ],
    "mnemonic": "jne",
    "opcode": "0x07",
    "operandTypes": Array [],
    "resultTypes": Array [],
  },
  Object {
    "argTypes": Array [],
    "mnemonic": "eq",
    "opcode": "0x08",
    "operandTypes": Array [
      "int32",
      "int32",
    ],
    "resultTypes": Array [
      "boolean",
    ],
  },
  Object {
    "argTypes": Array [],
    "mnemonic": "ne",
    "opcode": "0x09",
    "operandTypes": Array [
      "int32",
      "int32",
    ],
    "resultTypes": Array [
      "boolean",
    ],
  },
  Object {
    "argTypes": Array [],
    "mnemonic": "lt",
    "opcode": "0x0a",
    "operandTypes": Array [
      "int32",
      "int32",
    ],
    "resultTypes": Array [
      "boolean",
    ],
  },
  Object {
    "argTypes": Array [],
    "mnemonic": "gt",
    "opcode": "0x0b",
    "operandTypes": Array [
      "int32",
      "int32",
    ],
    "resultTypes": Array [
      "boolean",
    ],
  },
  Object {
    "argTypes": Array [],
    "mnemonic": "lte",
    "opcode": "0x0c",
    "operandTypes": Array [
      "int32",
      "int32",
    ],
    "resultTypes": Array [
      "boolean",
    ],
  },
  Object {
    "argTypes": Array [],
    "mnemonic": "gte",
    "opcode": "0x0d",
    "operandTypes": Array [
      "int32",
      "int32",
    ],
    "resultTypes": Array [
      "boolean",
    ],
  },
]
`;
